coderd/database/queries/workspaces.sql

```diff
@@ -161,17 +178,41 @@ WHERE
 			name ILIKE '%' || @name || '%'
 		ELSE true
 	END
+	-- Filter by agent status
+	-- has-agent: is only applicable for workspaces in "start" transition. Stopped and deleted workspaces don't have agents.
+	AND CASE
+		WHEN @has_agent :: text != '' THEN
+			latest_build.transition = 'start'::workspace_transition
+			AND CASE
+				WHEN @has_agent = 'timeout' THEN
+					latest_build.first_connected_at IS NULL AND (latest_build.created_at + latest_build.connection_timeout_seconds * INTERVAL '1 second' < NOW())
+				WHEN @has_agent = 'connecting' THEN
+					latest_build.first_connected_at IS NULL
+				WHEN @has_agent = 'disconnected' THEN
+					(
+						latest_build.disconnected_at IS NOT NULL AND
+						latest_build.disconnected_at > latest_build.last_connected_at
+					) OR (
+						latest_build.last_connected_at IS NOT NULL AND
+						latest_build.last_connected_at + INTERVAL '1 second' * @agent_inactive_disconnect_timeout :: bigint < NOW()
+					)
+				WHEN @has_agent = 'connected' THEN
+					latest_build.last_connected_at IS NOT NULL
```

Comment:
Should this case also verify that the state isn't actually `disconnected` either via timeout or `disonnected_at`?

Reply:
I extended the condition.

Frankly speaking,, I'm thinking about refactoring that part of the code and creating an extra "runtime" column for `agent status`. I'm not quite sure if it's possible to use SQL clauses to reflect this [logic](https://github.com/coder/coder/blob/main/coderd/workspaceagents.go#L724-L750). I really wouldn't like to create functions...

Reply:
I guess that you can take a look at the CASE logic implementation one more time as I've rewritten it to look similar to the Go code.

Reply:
@mtojek I think the case logic is exactly what you want. But you would have to add that case logic to all reads. So either remove all `GetWorkspaceAgentsByXXX` and make 1 with a filter (like workspaces/template/etc). Or you could also make a `view`. I looked into this before with sqlc in an issue here:

https://github.com/coder/coder/issues/2201

Reply:
Hey @Emyrk, thanks for jumping in!

I had a chat about this issue with @mafredri to try `WITH` expression first. I guess that I can transform it into a `VIEW` if there are performance/clean-sql concerns. Otherwise, I am happy to do this in a follow-up.

Reply:
@Emyrk @mafredri After further research I noticed a possible blocker for a view. The CASE logic 
 requires `agent_inactive_disconnect_timeout_seconds`, which is a coderd option, and it's currently provided via query.

Unless we can change it to a static value, then I suppose we can't use views here.

Reply:
That's a good point @mtojek, it definitely hampers the use case. Ideally this value would come from either the terraform provider or be stored in the database by some other means, at which point it can be part of the query.

I don't find it unthinkable that we would serialize certain coder server runtime properties in the database, either as a temp table or one that is (re)written on startup. The inactivity seconds could then be stored there and joined.